src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java

```diff
@@ -315,14 +351,60 @@ private void updateCISShadowReportedState(Map<String, Object> reportedState) {
         });
     }
 
-    private void publishToGetCISShadowTopic() {
+    private void reportShadowReceived() {
+        CompletableFuture<?> shadowReceived = this.shadowGetResponseReceived.get();
+        if (shadowReceived != null) {
+            shadowReceived.complete(null);
+        }
+    }
+
+    @SuppressWarnings("PMD.AvoidCatchingGenericException")
+    private void fetchCISShadowWithRetriesAsync() {
+        synchronized (getShadowLock) {
+            if (getShadowTask != null && !getShadowTask.isDone()) {
+                // operation already in progress
+                return;
+            }
+            getShadowTask = executorService.submit(() -> {
+                try {
+                    RetryUtils.runWithRetry(
+                            GET_CIS_SHADOW_RETRY_CONFIG,
+                            () -> {
+                                CompletableFuture<?> shadowReceived = this.shadowGetResponseReceived.updateAndGet(
+                                        ignore -> new CompletableFuture<>());
+                                publishToGetCISShadowTopic().get(mqttOperationTimeoutSeconds, TimeUnit.SECONDS);
+                                // await shadow get accepted, rejected
+                                shadowReceived.get(mqttOperationTimeoutSeconds, TimeUnit.SECONDS);
+                                return null;
+                            },
+                            "get-cis-shadow",
+                            LOGGER);
+                } catch (InterruptedException ignore) {
+                    Thread.currentThread().interrupt();
+                } catch (Exception e) {
+                    LOGGER.atError().cause(e).log("unable to get CIS shadow");
+                }
+            });
+        }
+    }
+
+    private void cancelGetCISShadow() {
+        synchronized (getShadowLock) {
+            if (getShadowTask != null) {
+                getShadowTask.cancel(true);
+            }
+        }
+    }
+
+    private CompletableFuture<Integer> publishToGetCISShadowTopic() {
         LOGGER.atDebug().log("Publishing to get shadow topic");
         GetShadowRequest getShadowRequest = new GetShadowRequest();
         getShadowRequest.thingName = shadowName;
-        iotShadowClient.PublishGetShadow(getShadowRequest, QualityOfService.AT_LEAST_ONCE).exceptionally(e -> {
-            LOGGER.atWarn().cause(e).log("Unable to retrieve CIS shadow");
-            return null;
-        });
+        return iotShadowClient.PublishGetShadow(getShadowRequest, QualityOfService.AT_LEAST_ONCE)
```

Comment:
use qos 0. Otherwise when we're offline this is going to fill up the spooler.

you may also want to not retry when we know that we're offline.

Reply:
good point. and yeah when cda goes offline the task will be cancelled, but i didn't handle case if we start offline will fix